src/dma.c

```diff
@@ -316,45 +398,64 @@ void dma_read(void *ram_address, unsigned long pi_address, unsigned long len)
  *
  * @param[in] ram_address
  *            Pointer to a buffer to read data from
- * @param[in] pi_address
- *            Memory address of the peripheral to write to
+ * @param[in] rom_address
+ *            Cartridge address to write to (must be in range (0x10000000-0x1FFFFFFF).
  * @param[in] len
  *            Length in bytes to write to peripheral
+ * 
+ * @note This function has always had an historical mistake: the pi_address is mangled
+ *       to be forced into the ROM area (0x10000000-0x1FFFFFFF). This is wrong as the
+ *       PI bus has full 32-bit address, and the same function could have been used
+ *       to access the whole range.
+ *       If you need to read outside the ROM area, use #dma_write_raw_async instead.
  */
-void dma_write(const void * ram_address, unsigned long pi_address, unsigned long len) 
+void dma_write(const void * ram_address, unsigned long rom_address, unsigned long len) 
 {
-    assert(len > 0);
-
-    dma_write_raw_async(ram_address, pi_address, len);
+    rom_address = (rom_address | 0x10000000) & 0x1FFFFFFF;
+    dma_write_raw_async(ram_address, rom_address, len);
     dma_wait();
 }
 
 /**
- * @brief Read a 32 bit integer from a peripheral
+ * @brief Read a 32 bit integer from a peripheral using the CPU.
  *
  * @param[in] pi_address
  *            Memory address of the peripheral to read from
  *
  * @return The 32 bit value read from the peripheral
+ * 
+ * @note This function only works if the specified PI address falls within a range
+ *       which is memory mapped on the CPU. See #io_accessible for more information.
```

Comment:
Maybe we can assert if it is accessible? Would it be too much a hit for performance on a debug build?

Reply:
I preferred not to for performance reasons. There are rationales here:

 * Currently there are no known PI peripherals that operate outside of the accessible areas. I don't know what PicoCart64 is going to do for instance, but I guess they'd rather fall within the range for flexibility. So `pi_accessible` is mostly for documentation purposes, and checking `pi_accessible` in `dma_read_async` is just... uhm.. pedantry. The assert in io_read would never trigger (if not for programmer bugs, that would trigger anyway in other forms).
 * Normally `io_read` is called with some "fixed" / "hardcoded" address. Outside of library code, It is very uncommon to call it with an address which is "unknown" at compile time. So the address would either be correct (and the code will work) or wrong because eg a typo during development (and the code will crash/miswork right away). It doesn't seem an assert is adding much here.
 * We don't have a real story for non-debug builds, especially for disabling NDEBUG in libdragon. One has to recompile everything and know what to do. n64.mk doesn't help at the moment. I wouldn't be surprised if most homebrew is released with NDEBUG on.

So it's a bit on the edge, but the check is a bit slow compared to io_read itself which is a memory op, and I don't think it's helping much.

Reply:
I agree, except I use NDEBUG for my final builds :)